File: backend/app/services/booking_operations.py
"""Lógica operativa de reservas — espejo en Python de
`frontend/src/features/admin/lib/booking-operations.ts`.

Una reserva de transporte se descompone en SERVICIOS OPERATIVOS (piernas):
- LLEGADA  (route 'airport-hotel'): Aeropuerto → Hotel.
- SALIDA   (route 'hotel-airport'): Hotel → Aeropuerto.
- ROUND TRIP (trip_type 'roundtrip'): genera AMBAS piernas.

Para la operación de la empresa, en cada pierna es IMPRESCINDIBLE:
- de dónde sale y a dónde va (origen → destino),
- la HORA DEL SERVICIO. En las salidas, la hora del servicio es el **pickup
  time** (cuándo recogemos al pasajero en el hotel), que por defecto es 3 horas
  antes de la hora del vuelo de salida.

Mantener esta lógica idéntica a la del frontend para que el PDF muestre lo mismo
que ve el admin.
"""

from __future__ import annotations

_AIRPORT_FALLBACK = "Aeropuerto"
_HOTEL_FALLBACK = "Hotel / Pickup"


def _clean_date(value: str | None) -> str:
    if not value:
        return ""
    return value[:10]


def _minutes_from_time(value: str | None) -> int | None:
    if not value or len(value) < 4 or ":" not in value:
        return None
    try:
        hours, minutes = value[:5].split(":")
        return int(hours) * 60 + int(minutes)
    except (ValueError, TypeError):
        return None


def _time_from_minutes(value: int) -> str:
    return f"{value // 60:02d}:{value % 60:02d}"


def subtract_three_hours(value: str | None) -> str | None:
    """Pickup por defecto = hora del vuelo de salida − 3 horas."""
    minutes = _minutes_from_time(value)
    if minutes is None:
        return None
    return _time_from_minutes(max(0, minutes - 180))


def _departure_date_from_metadata(booking: dict) -> str:
    meta = booking.get("metadata")
    if isinstance(meta, dict):
        raw = meta.get("departureDate")
        if isinstance(raw, str):
            return _clean_date(raw)
    return ""


def build_operation_legs(booking: dict) -> list[dict]:
    """Descompone una reserva (dict de `_booking_to_dict`) en sus piernas.

    Cada pierna devuelta:
        type:         'arrival' | 'departure' | 'service'
        label:        'LLEGADA' | 'SALIDA' | 'SERVICIO'
        service_date: 'YYYY-MM-DD'
        service_time: hora del servicio (en salidas = pickup time)
        origin:       de dónde sale
        destination:  a dónde va
        flight:       número de vuelo
        airline:      aerolínea
        flight_time:  hora del vuelo (referencia; en salidas distinta al pickup)
    """
    legs: list[dict] = []

    trip_type = (booking.get("trip_type") or "").lower()
    route = booking.get("route")
    is_roundtrip = trip_type == "roundtrip"
    arrival_date = _clean_date(booking.get("booking_date"))
    departure_date = _departure_date_from_metadata(booking) or arrival_date

    pickup = booking.get("pickup_location")
    dropoff = booking.get("dropoff_location")

    has_arrival_data = any(
        booking.get(k) for k in ("flight_number", "arrival_time", "arrival_airline")
    )
    has_departure_data = any(
        booking.get(k)
        for k in ("departure_flight_number", "departure_time", "departure_airline", "pickup_time")
    )

    # ─── LLEGADA ───
    want_arrival = route == "airport-hotel" or is_roundtrip or (route is None and has_arrival_data)
    if want_arrival:
        legs.append(
            {
                "type": "arrival",
                "label": "LLEGADA",
                "service_date": arrival_date,
                "service_time": booking.get("booking_time") or booking.get("arrival_time") or "",
                "origin": pickup or _AIRPORT_FALLBACK,
                "destination": dropoff or _HOTEL_FALLBACK,
                "flight": booking.get("flight_number") or "",
                "airline": booking.get("arrival_airline") or "",
                "flight_time": booking.get("arrival_time") or "",
            }
        )

    # ─── SALIDA ───
    departure_service_time = (
        booking.get("pickup_time")
        or subtract_three_hours(booking.get("departure_time"))
        or booking.get("departure_time")
        or ""
    )
    if is_roundtrip:
        # En un round trip el pickup de ida es el aeropuerto y el drop es el hotel;
        # la salida es la ruta inversa: Hotel → Aeropuerto.
        legs.append(
            {
                "type": "departure",
                "label": "SALIDA",
                "service_date": departure_date,
                "service_time": departure_service_time,
                "origin": dropoff or _HOTEL_FALLBACK,
                "destination": pickup or _AIRPORT_FALLBACK,
                "flight": booking.get("departure_flight_number") or "",
                "airline": booking.get("departure_airline") or "",
                "flight_time": booking.get("departure_time") or "",
            }
        )
    elif route == "hotel-airport" or (route is None and has_departure_data):
        # Salida de una sola vía: el pickup es el hotel, el drop es el aeropuerto.
        legs.append(
            {
                "type": "departure",
                "label": "SALIDA",
                "service_date": arrival_date,
                "service_time": departure_service_time,
                "origin": pickup or _HOTEL_FALLBACK,
                "destination": dropoff or _AIRPORT_FALLBACK,
                "flight": booking.get("departure_flight_number")
                or booking.get("flight_number")
                or "",
                "airline": booking.get("departure_airline") or "",
                "flight_time": booking.get("departure_time") or "",
            }
        )

    # ─── Servicio genérico (sin ruta ni vuelos) ───
    if not legs:
        legs.append(
            {
                "type": "service",
                "label": "SERVICIO",
                "service_date": arrival_date,
                "service_time": (
                    booking.get("booking_time")
                    or booking.get("arrival_time")
                    or booking.get("departure_time")
                    or ""
                ),
                "origin": pickup or "",
                "destination": dropoff or "",
                "flight": booking.get("flight_number")
                or booking.get("departure_flight_number")
                or "",
                "airline": booking.get("arrival_airline") or booking.get("departure_airline") or "",
                "flight_time": "",
            }
        )

    return legs

File: frontend/src/features/admin/lib/booking-operations.ts
export type AdminOperationBooking = {
  id?: string;
  bookingDate: string;
  bookingTime?: string | null;
  pickupTime?: string | null;
  arrivalTime?: string | null;
  departureTime?: string | null;
  pickupLocation?: string | null;
  dropoffLocation?: string | null;
  flightNumber?: string | null;
  arrivalAirline?: string | null;
  departureFlightNumber?: string | null;
  departureAirline?: string | null;
  route?: string | null;
  tripType?: string | null;
  notes?: string | null;
  metadata?: Record<string, unknown> | null;
  customer?: {
    name?: string | null;
    email?: string | null;
    phone?: string | null;
  } | null;
  passengers?: number;
  status?: string;
};

export type OperationType = 'arrival' | 'departure' | 'other';

export type AdminOperationEvent<T extends AdminOperationBooking = AdminOperationBooking> = {
  key: string;
  booking: T;
  operationType: OperationType;
  serviceDate: string;
  serviceTime: string;
  flight: string;
  airline: string;
  hotel: string;
  routeLabel: string;
  origin: string;
  destination: string;
};

// Mismos fallbacks que `booking_operations.py` (backend) — deben coincidir.
const AIRPORT_FALLBACK = 'Aeropuerto';
const HOTEL_FALLBACK = 'Hotel / Pickup';

function cleanDate(value?: string | null) {
  if (!value) return '';
  if (value.includes('T')) return value.slice(0, 10);
  return value.slice(0, 10);
}

function minutesFromTime(value?: string | null) {
  if (!value || !/^\d{1,2}:\d{2}$/.test(value)) return null;
  const [hours, minutes] = value.split(':').map(Number);
  if (Number.isNaN(hours) || Number.isNaN(minutes)) return null;
  return hours * 60 + minutes;
}

function timeFromMinutes(value: number) {
  const hours = Math.floor(value / 60);
  const minutes = value % 60;
  return `${String(hours).padStart(2, '0')}:${String(minutes).padStart(2, '0')}`;
}

function subtractThreeHours(value?: string | null) {
  const minutes = minutesFromTime(value);
  if (minutes === null) return null;
  return timeFromMinutes(Math.max(0, minutes - 180));
}

function departureDateFromMetadata(booking: AdminOperationBooking) {
  const raw = booking.metadata && typeof booking.metadata === 'object'
    ? booking.metadata.departureDate
    : null;

  if (typeof raw !== 'string') return '';
  return cleanDate(raw);
}

export function getOperationType(booking: AdminOperationBooking): OperationType {
  if (booking.route === 'airport-hotel') return 'arrival';
  if (booking.route === 'hotel-airport') return 'departure';
  return 'other';
}

export function getOperationBadge(input: AdminOperationBooking | AdminOperationEvent) {
  const type = 'operationType' in input ? input.operationType : getOperationType(input);
  if (type === 'arrival') return { label: 'LLEGADA', className: 'bg-blue-500/15 text-blue-300' };
  if (type === 'departure') return { label: 'SALIDA', className: 'bg-orange-100 text-orange-700' };
  return { label: 'SERVICIO', className: 'bg-white/10 text-white/60' };
}

export function getOperationHotel(input: AdminOperationBooking | AdminOperationEvent) {
  if ('hotel' in input) return input.hotel;
  return getOperationType(input) === 'departure'
    ? input.pickupLocation || 'Pickup pending'
    : input.dropoffLocation || 'Destination pending';
}

export function getOperationFlight(input: AdminOperationBooking | AdminOperationEvent) {
  if ('flight' in input) return input.flight;
  return getOperationType(input) === 'departure'
    ? input.departureFlightNumber || input.flightNumber || '---'
    : input.flightNumber || input.departureFlightNumber || '---';
}

export function getOperationTime(input: AdminOperationBooking | AdminOperationEvent) {
  if ('serviceTime' in input) return input.serviceTime;

  if (getOperationType(input) === 'departure') {
    return input.pickupTime || subtractThreeHours(input.departureTime) || input.departureTime || input.bookingTime || '99:99';
  }

  return input.bookingTime || input.arrivalTime || input.departureTime || '99:99';
}

function operationPriority(input: AdminOperationBooking | AdminOperationEvent) {
  const type = 'operationType' in input ? input.operationType : getOperationType(input);
  if (type === 'arrival') return 0;
  if (type === 'departure') return 1;
  return 2;
}

export function compareOperationBookings<T extends AdminOperationBooking | AdminOperationEvent>(a: T, b: T) {
  const dateA = 'serviceDate' in a ? a.serviceDate : cleanDate(a.bookingDate);
  const dateB = 'serviceDate' in b ? b.serviceDate : cleanDate(b.bookingDate);
  const dateDiff = dateA.localeCompare(dateB);
  if (dateDiff !== 0) return dateDiff;

  const opDiff = operationPriority(a) - operationPriority(b);
  if (opDiff !== 0) return opDiff;

  return getOperationTime(a).localeCompare(getOperationTime(b));
}

export function expandBookingOperations<T extends AdminOperationBooking>(booking: T): AdminOperationEvent<T>[] {
  const events: AdminOperationEvent<T>[] = [];
  const arrivalDate = cleanDate(booking.bookingDate);
  const departureDate = departureDateFromMetadata(booking);
  const isRoundTrip = booking.tripType === 'roundtrip';

  if (booking.route === 'airport-hotel' || isRoundTrip) {
    events.push({
      key: `${booking.id || arrivalDate}-arrival`,
      booking,
      operationType: 'arrival',
      serviceDate: arrivalDate,
      serviceTime: booking.bookingTime || booking.arrivalTime || '99:99',
      flight: booking.flightNumber || '---',
      airline: booking.arrivalAirline || '',
      hotel: booking.dropoffLocation || 'Hotel pending',
      routeLabel: 'Airport -> Hotel',
      origin: booking.pickupLocation || AIRPORT_FALLBACK,
      destination: booking.dropoffLocation || HOTEL_FALLBACK,
    });
  }

  if (booking.route === 'hotel-airport') {
    events.push({
      key: `${booking.id || arrivalDate}-departure`,
      booking,
      operationType: 'departure',
      serviceDate: arrivalDate,
      serviceTime: booking.pickupTime || subtractThreeHours(booking.departureTime) || booking.departureTime || booking.bookingTime || '99:99',
      flight: booking.departureFlightNumber || booking.flightNumber || '---',
      airline: booking.departureAirline || '',
      hotel: booking.pickupLocation || 'Pickup pending',
      routeLabel: 'Hotel -> Airport',
      origin: booking.pickupLocation || HOTEL_FALLBACK,
      destination: booking.dropoffLocation || AIRPORT_FALLBACK,
    });
  } else if (isRoundTrip) {
    // Si la reserva no trae metadata.departureDate (reservas viejas creadas
    // antes de este fix, o el admin no la llenó), cae a arrivalDate — igual
    // que build_operation_legs() en el backend — para que la pierna de
    // SALIDA siempre aparezca, aunque sea con la fecha equivocada en vez de
    // desaparecer por completo.
    const effectiveDepartureDate = departureDate || arrivalDate;
    events.push({
      key: `${booking.id || effectiveDepartureDate}-departure`,
      booking,
      operationType: 'departure',
      serviceDate: effectiveDepartureDate,
      serviceTime: booking.pickupTime || subtractThreeHours(booking.departureTime) || booking.departureTime || '99:99',
      flight: booking.departureFlightNumber || '---',
      airline: booking.departureAirline || '',
      hotel: booking.dropoffLocation || booking.pickupLocation || 'Hotel pending',
      routeLabel: 'Hotel -> Airport',
      // Round trip: la salida es la ruta inversa de la llegada (hotel -> aeropuerto).
      origin: booking.dropoffLocation || HOTEL_FALLBACK,
      destination: booking.pickupLocation || AIRPORT_FALLBACK,
    });
  }

  if (events.length === 0) {
    events.push({
      key: `${booking.id || arrivalDate}-service`,
      booking,
      operationType: 'other',
      serviceDate: arrivalDate,
      serviceTime: booking.bookingTime || booking.arrivalTime || booking.departureTime || '99:99',
      flight: booking.flightNumber || booking.departureFlightNumber || '---',
      airline: booking.arrivalAirline || booking.departureAirline || '',
      hotel: booking.dropoffLocation || booking.pickupLocation || 'Pending',
      routeLabel: 'Service',
      origin: booking.pickupLocation || '',
      destination: booking.dropoffLocation || '',
    });
  }

  return events.sort(compareOperationBookings);
}
